tools/rewriter/GenCallAsm.cpp

```diff
@@ -391,27 +409,31 @@ static void emit_copy_args(AsmWriter &aw, size_t stack_return_size, size_t stack
     }
 
     // Copy stack args to target stack
-    if (stack_arg_count > 0) {
+    if (stack_arg_size > 0) {
       // Set rax to the caller's stack so we can copy the stack args to the
       // compartment's stack.
       add_comment_line(
           aw, "Copy stack arguments from the caller's stack to the compartment");
       // Load addr of top of caller compartment's stack into rax, clobbering r12
       std::string expr = emit_load_sp_offset(aw, caller_pkey, "r12"s);
+      add_asm_line(aw, "subq $" + std::to_string(stack_arg_size + stack_arg_padding) + ", %rsp");
       add_asm_line(aw, "movq %" + expr + ", %rax");
-      // This is effectively a memcpy of size `stack_arg_size` from the caller's
-      // stack to the compartment's
-      for (int i = 0; i < stack_arg_size; i += 8) {
-        // We must take the preserved registers we pushed on the caller's stack
-        // into account (including rbp) when determining the location of the stack
-        // args
-        size_t caller_stack_size =
-            stack_arg_size + ((preserved_registers.size() + 1) * 8);
-        // The index into the caller's stack is backwards since pushq will copy to
-        // the compartment's stack from the highest addresses to the lowest.
-        add_asm_line(aw,
-                     "pushq " + std::to_string(caller_stack_size - i) + "(%rax)");
-      }
+      // We must take the preserved registers we pushed on the caller's stack
+      // into account (including rbp and the return address) when determining
+      // the location of the stack args.
+      size_t offset = (preserved_registers.size() + 2) * 8;
+      emit_memcpy(aw, stack_arg_size, "rsp", "rax", "r10", offset);
```

Comment:
Don't split up construction and use of `expr`.

Is it preferable to use `mov`s and a separate modification of `rsp` here as opposed to the `push`es used before, or is there another reason to change this?

If we do use `mov` and a temporary, we should probably reuse `r12` which we just clobbered.

Reply:
I originally thought we would copy the exact size, not a padded to 8 bytes size. I can revert this bit.

Reply:
Reverted back to `pushq`s.